src/main/java/com/microsoft/aad/msal4j/HttpHelper.java

```diff
@@ -42,17 +73,79 @@ static IHttpResponse executeHttpRequest(final HttpRequest httpRequest,
                 HttpHelper.verifyReturnedCorrelationId(httpRequest, httpResponse);
             }
         }
+        processThrottlingInstructions(httpResponse, requestContext);
+
+        return httpResponse;
+    }
+
+    private static IHttpResponse executeHttpRequestWithRetries(HttpRequest httpRequest, IHttpClient httpClient)
+            throws Exception {
+        IHttpResponse httpResponse = null;
+        for (int i = 0; i < RETRY_NUM; i++) {
+            httpResponse = httpClient.send(httpRequest);
+            if (!isRetriable(httpResponse)) {
+                break;
+            }
+            Thread.sleep(RETRY_DELAY_MS);
+        }
         return httpResponse;
     }
 
-    private static void addRequestInfoToTelemetry(final HttpRequest httpRequest, HttpEvent httpEvent){
-        try{
+    private static void checkForThrottling(RequestContext requestContext) {
+        if (requestContext.clientApplication() instanceof PublicClientApplication &&
+                requestContext.apiParameters() != null) {
+            String requestThumbprint = getRequestThumbprint(requestContext);
+
+            long retryInMs = ThrottlingCache.retryInMs(requestThumbprint);
+
+            if (retryInMs > 0) {
+                throw new MsalThrottlingException(retryInMs);
```

Comment:
I was under the impression that throttling concept is not exposed to client. We should just throw the old exception, as if the client had made the HTTP call and the server returned the same error. 

I think this makes sense because it would behave well with retry policies that customers have added on their own.

Reply:
Not sure about it, if we get instructions from STS to throttle some requests, we do it and do it transparently for app develpers.

Reply:
I sent an email to Sasha for clarifications as I am also not sure about this.